src/core/lib/channel/channelz.cc

```diff
@@ -277,62 +280,64 @@ grpc_json* ServerNode::RenderJson() {
   return top_level_json;
 }
 
-SocketAddress::SocketAddress() : type_(AddressType::kUnset), port_(-1) {}
-
-SocketAddress::SocketAddress(SocketAddress&& other) {
-  type_ = other.type_;
-  port_ = other.port_;
-  blob_.reset((other.blob_.release()));
-}
-
-void SocketAddress::PopulateJson(const char* name, grpc_json* json) {
-  if (type_ == AddressType::kUnset) {
-    return;
-  }
+static void PopulateSocketAddressJson(grpc_json* json, const char* name,
+                                      const char* addr_str) {
+  if (addr_str == nullptr) return;
   grpc_json* json_iterator = nullptr;
   json_iterator = grpc_json_create_child(json_iterator, json, name, nullptr,
                                          GRPC_JSON_OBJECT, false);
   json = json_iterator;
   json_iterator = nullptr;
-  switch (type_) {
-    case AddressType::kTcpAddress:
-      json_iterator =
-          grpc_json_create_child(json_iterator, json, "tcpip_address", nullptr,
-                                 GRPC_JSON_OBJECT, false);
-      json = json_iterator;
-      json_iterator = nullptr;
-      json_iterator =
-          grpc_json_add_number_string_child(json, json_iterator, "port", port_);
-      json_iterator =
-          grpc_json_create_child(json_iterator, json, "ip_address", blob_.get(),
-                                 GRPC_JSON_STRING, false);
-      break;
-    case AddressType::kUdsAddress:
+  int port_num = -1;
+  grpc_uri* uri = grpc_uri_parse(addr_str, true);
+  if (uri != nullptr && (strcmp(uri->scheme, "fd") != 0)) {
+    const char* host_port = uri->path;
+    if (*host_port == '/') ++host_port;
+    if (strcmp(uri->scheme, "unix") == 0) {
       json_iterator = grpc_json_create_child(json_iterator, json, "uds_address",
                                              nullptr, GRPC_JSON_OBJECT, false);
       json = json_iterator;
       json_iterator = nullptr;
       json_iterator =
-          grpc_json_create_child(json_iterator, json, "filename", blob_.get(),
-                                 GRPC_JSON_STRING, false);
-      break;
-    case AddressType::kDirectChannelAddress:
+          grpc_json_create_child(json_iterator, json, "filename",
+                                 gpr_strdup(host_port), GRPC_JSON_STRING, true);
+    } else {
+      char* host = nullptr;
+      char* port = nullptr;
+      if (strcmp(uri->scheme, "localhost") == 0) {
```

Comment:
This seems wrong.  Why would the scheme be "localhost"?

I suspect this needs to move down after the call to `gpr_split_host_port()`, and then check `host`.

Reply:
This was only coming from the string passed to the listening socket... I suspect that I can do some substitution earlier in the code

Reply:
To be a little more clear, this is actually coming from the listening socket's peer string, which is "localhost:21802"

When that gets parsed by the URI parser, we get `uri->scheme="localhost", uri->path="21802"`, which is not correct. Do you think I should fix this up in the uri parser? Or do an explicit check for localhost before even passing to the uri parser

Reply:
Yeah, I guess the listen socket's address string won't necessarily be a URI.  The string passed to `grpc_server_add_insecure_http2_port()` will get passed directly to `grpc_blocking_resolve_address()` [here](https://github.com/grpc/grpc/blob/master/src/core/ext/transport/chttp2/server/chttp2_server.cc#L311), with no URI parsing happening first.  We do have a [horrible hack](https://github.com/grpc/grpc/blob/master/src/core/lib/iomgr/resolve_address_posix.cc#L57) in place to support `unix:` URIs, but we have a bug somewhere (can't find it at the moment) to fix that.

For now, my suggestion would be to first parse it as a URI and then check whether the scheme is one of the 3 specific ones we know how to handle (`ipv4`, `ipv6`, and `unix`).  If it's none of those, then even if it parsed as a URI, we should ignore that and just report the full string as an `OtherAddress`.

Reply:
Ok, done. I will look into adding an API to endpoint to grab the local and have that out soon